doc/sonic-build-system/SONiC-OpenSSL-FIPS-140-3.md

```diff
@@ -0,0 +1,140 @@
+| -------------------- | --------------- | -------------------------------------------- |
+| OpenSSL              | Python, OpenSSH | Cyptography and SSL/TLS ToolKit              |
+| Kerberos             | OpenSSH         | Kerboros contains builtin crypto module      |
+| Golang               | sonic-restapi   | Golang contains builtin crypt module         |
+| Libgcrypto           | GPG             | A general purpose cryptographic library originally based on code from GnuPG |
+| Kernel Crypto        | --              | Linux crypto kernel module |
+
+In Scopes:
+1. OpenSSL
+2. Kerberos
+3. Golang
+
+Out of Scopes:
+1. Linux Kernel
+2. Libgcrypt
+
+## SONiC FIPS Configuration
+### Enable FIPS on system level
+Add the Linux System parameter fips=1, in grub config, one of implemetation as below:
+
+cat /etc/grub.d/99-fips.cfg
+```
+GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT fips=1"
+```
+
+To validate the FIPS enabled, grep 'fips=1' /proc/cmdline.
+
+### Enable FIPS on application level
+```
+export ENABLE_FIPS=1
+```
+
+Alternative option for the golang applications only:
+```
+export GOLANG_FIPS=1
+```
+
+Alternative option for the OpenSSL applications only:
+
+see https://www.openssl.org/docs/manmaster/man7/openssl-env.html
+```
+export OPENSSL_CONFIG=/usr/lib/ssl/openssl-fips.cnf
+```
+
+## OpenSSL FIPS 140-3
+
+![FIPS Overview!](images/fips-overview.png)
+
+### OpenSSL Engine
+OpenSSL supports engine cryptographic modules in the form of engine objects, and provides a reference-counted mechanism to allow them to be dynamically loaded in and out of the running application. An engine object can implement one or all cryptographic algorithms.
+
+### SymCrypt OpenSSL Engine
+The [SymCrypt engine for OpenSSL (SCOSSL)](https://github.com/microsoft/SymCrypt-OpenSSL) allows the use of OpenSSL with [SymCrypt](https://github.com/microsoft/SymCrypt) as the provider for core cryptographic operations. It leverages the OpenSSL engine interface to override the cryptographic implementations in OpenSSL's libcrypto. The primary motivation for this is to support FIPS certification, as vanilla OpenSSL 1.1.1 does not have a FIPS-certified cryptographic module. Microsoft will submit the FIPS 140-3 reports for SymCrypt to CMVP.
+
+The SymCrypt Engine is one of the implementation to support FIPS, The [wolfSSL engine](https://github.com/wolfSSL/wolfEngine) is another option.
+
+### OpenSSL configuration for SymCrypt Engine
+
+/usr/lib/ssl/openssl-fips.cnf 
+```
+openssl_conf = openssl_init
+[ openssl_init ]
+engines = engine_section
+
+[ engine_section ]
+symcrypt = symcrypt_section
+
+[ symcrypt_section ]
+engine_id = symcrypt
+dynamic_path = /usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+default_algorithms = ALL
+```
+
+### OpenSSL configuration enhancement
+When fips=1 is set in /proc/cmdline, the OpenSSL default config file is changed to "/usr/lib/ssl/openssl-fips.cnf".
```

Comment:
>the OpenSSL default config file is changed [](http://example.com/codeflow?start=37&length=42)

Just curios what is the config file for fips=0?

Reply:
When fips not set to 1, the other one /usr/lib/ssl/openssl.cnf will be used.